sql/hive/src/main/scala/org/apache/spark/sql/hive/execution/InsertIntoHiveTable.scala

```diff
@@ -230,12 +255,28 @@ case class InsertIntoHiveTable(
         var doHiveOverwrite = overwrite
 
         if (oldPart.isEmpty || !ifPartitionNotExists) {
+          // SPARK-29295: When insert overwrite to a Hive external table partition, if the
+          // partition does not exist, Hive will not check if the external partition directory
+          // exists or not before copying files. So if users drop the partition, and then do
+          // insert overwrite to the same partition, the partition will have both old and new
+          // data.
+          val updatedPart = if (overwrite && table.tableType == CatalogTableType.EXTERNAL) {
+            AlterTableAddPartitionCommand(
+              table.identifier, Seq((partitionSpec, None)), ifNotExists = true).run(sparkSession)
+            externalCatalog.getPartitionOption(
+              table.database,
+              table.identifier.table,
+              partitionSpec)
+          } else {
+            oldPart
```

Comment:
we should compute `oldPart` here, so that we don't need to compute it when overwriting external table.

Reply:
We need to know partition exists or not because if `ifPartitionNotExists` is true and partition exists, this command is no-op.

But we can add a condition when overwriting external table. Only when `oldPart` is empty, we need to call `AlterTableAddPartitionCommand`.

Reply:
Then we do no call `getPartitionOption` twice.